File: terry/programmers/stack-queue/printer.cpp
#include <string>
#include <queue>
#include <vector>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    
    priority_queue<int> pq;
    queue<int> q;
    
    for(int i = 0; i < priorities.size(); i++) {
        pq.push(priorities[i]);
        q.push(i);
    }
    
    while(q.size()) {
        answer++;
        while(pq.top() != priorities[q.front()]) {
            q.push(q.front());
            q.pop();
        }
        pq.pop();
        if(q.front() == location) {
            return answer;
        }
        q.pop();
    }
    
    return answer;
}

File: taehwa/programmers/stack-queue/printer.ts
export default function solution(priorities: number[], location: number) {
  const queue = priorities.map((priority, index) => [priority, index]);

  let isPrinted = false;
  let counter = 0;

  while (!isPrinted) {
    const head = queue.shift() as [number, number];
    const [priority] = head;

    if (queue.some((element) => element[0] > priority)) {
      queue.push(head);
    } else {
      counter++;
      if (head[1] === location) {
        isPrinted = true;
      }
    }
  }

  return counter;
}
